x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -1337,25 +1340,43 @@ public void cleanUpFeature(
             )
         );
 
-        Map<String, Boolean> results = new ConcurrentHashMap<>();
+        ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> cleanedUpIndicesListener = ActionListener.wrap(
+            success -> {
+                if (memoryTracker.get() != null) {
+                    memoryTracker.get().awaitAndClear(ActionListener.wrap(
+                        cacheCleared -> unsetResetModeListener.onResponse(success),
+                        clearFailed -> {
+                            logger.error("failed to clear memory tracker cache via machine learning reset feature API", clearFailed);
+                            unsetResetModeListener.onResponse(success);
+                        }
+                    ));
+                    return;
+                }
+                unsetResetModeListener.onResponse(success);
+            },
+            failure -> {
+                logger.error("failed to clear .ml-* indices via reset feature API", failure);
+                unsetResetModeListener.onFailure(failure);
+            }
+        );
 
         ActionListener<ListTasksResponse> afterWaitingForTasks = ActionListener.wrap(
             listTasksResponse -> {
                 listTasksResponse.rethrowFailures("Waiting for indexing requests for .ml-* indices");
                 if (results.values().stream().allMatch(b -> b)) {
-                    // Call into the original listener to clean up the indices
-                    SystemIndexPlugin.super.cleanUpFeature(clusterService, client, unsetResetModeListener);
+                    // Call into the original listener to clean up the indices and then clear ml memory cache
+                    SystemIndexPlugin.super.cleanUpFeature(clusterService, client, cleanedUpIndicesListener);
```

Comment:
Since the memory tracker cleanup waits for refreshes to finish, I would do this index cleanup after clearing the memory tracker.  It should avoid logging of spurious errors from refreshes that fail because the indices they're accessing get deleted.

Or was there a good reason for clearing the memory tracker last?

Reply:
@droberts195, I figured if the job's potentially still existed, it would be good to keep around their estimates. But, since all jobs should be closed by this point, clearing the tracker earlier is probably ok.

Reply:
Yes, the jobs should be closed, and banned from reopening by the reset-in-progress cluster setting.  I think flipping the order may avoid log spam.